libs-haskell/da-hs-language-server/src/DA/LanguageServer/Server.hs

```diff
@@ -46,26 +59,143 @@ runServer loggerH reqHandler notifHandler notifChan = do
     -- order not to mess up the output too much. Verified that this breaks
     -- the language server tests without the redirection.
     putStr " " >> hFlush stdout
+    clientMsgChan <- newTChanIO
+    void $ LSP.runWithHandles
+        stdin
+        newStdout
+        (const $ Right (), handleInit clientMsgChan)
+        (handlers clientMsgChan)
+        options
+        Nothing
+    where
+        reqHandler' :: (ServerRequest, LspId) -> IO LSP.FromServerMessage
+        reqHandler' (req, reqId) =
+            reqHandler
+                (\res -> ResponseMessage "2.0" (responseId reqId) (Just res) Nothing)
+                (\err -> ResponseMessage "2.0" (responseId reqId) Nothing (Just $ ResponseError err "" Nothing))
+                req
+        handleInit :: TChan LSP.FromClientMessage -> LSP.LspFuncs () -> IO (Maybe LSP.ResponseError)
+        handleInit clientMsgChan LSP.LspFuncs{..} = do
+            _ <- forkIO $ forever $ do
```

Comment:
These forkIO calls have a nasty smell to them... Are they temporary or permanent? Do we need better lifetime control on this function?

Reply:
Good point, I’ve changed it to tear down all of `runServer` if any of these threads die which corresponds to what the `runServer` in `json-rpc` does.